timemachine/fe/rbfe.py

```diff
@@ -459,6 +486,66 @@ def estimate_free_energy_given_initial_states(initial_states, protocol, temperat
     )
 
 
+def _optimize_coords_along_states(initial_states):
+    # use the end-state to define the optimization settings
+    end_state = initial_states[0]
+    ligand_coords = end_state.x0[end_state.ligand_idxs]
+    r_i = np.expand_dims(ligand_coords, axis=1)
+    r_j = np.expand_dims(end_state.x0, axis=0)
+    d_ij = np.linalg.norm(jax_utils.delta_r(r_i, r_j, box=end_state.box0), axis=-1)
+    cutoff = 0.5  # in nanometers
+    free_idxs = np.where(np.any(d_ij < cutoff, axis=0))[0].tolist()
+    x_opt = end_state.x0
+    x_traj = []
+    for initial_state in initial_states:
+        bound_impls = [p.bound_impl(np.float32) for p in initial_state.potentials]
+        val_and_grad_fn = minimizer.get_val_and_grad_fn(bound_impls, initial_state.box0, initial_state.lamb)
+        assert np.all(np.isfinite(x_opt)), "Initial coordinates contain nan or inf"
+        x_opt = minimizer.local_minimize(x_opt, val_and_grad_fn, free_idxs)
+        x_traj.append(x_opt)
+        assert np.all(np.isfinite(x_opt)), "Minimization resulted in a nan"
+        del bound_impls
+
+    return x_traj
+
+
+def optimize_coordinates(initial_states):
+    """
+    Optimize geometries of the initial states. The initial states are modified in place.
+    """
+    all_xs = []
+    lambda_schedule = np.array([s.lamb for s in initial_states])
+
+    # check for monotonic, any subsequence of a monotonic sequence is also monotonic.
+    assert np.all(np.diff(lambda_schedule) > 0)
+
+    lhs_initial_states = []
+    rhs_initial_states = []
+
+    for state in initial_states:
+        if state.lamb < 0.5:
+            lhs_initial_states.append(state)
+        else:
+            rhs_initial_states.append(state)
+
+    # go from lambda 0 -> 0.5
+    if len(lhs_initial_states) > 0:
+        lhs_xs = _optimize_coords_along_states(lhs_initial_states)
+        for xs in lhs_xs:
+            all_xs.append(xs)
+
+    # go from lambda 1 -> 0.5 and reverse the coordinate trajectory and lambda schedule
+    if len(rhs_initial_states) > 0:
+        rhs_xs = _optimize_coords_along_states(rhs_initial_states[::-1])[::-1]
+        for xs in rhs_xs:
+            all_xs.append(xs)
+
+    for state, coords in zip(initial_states, all_xs):
+        # sanity check that no atom has moved more than 7 angstroms away (arbitrary)
+        assert np.amax(np.linalg.norm(state.x0 - coords, axis=1)) < 0.7
```

Comment:
Should "close to initial coords" appear only as an assertion after optimization, or should this also be part of the optimization objective?

(If the optimization objective allows to go arbitrarily far away from initial coords, then an assertion afterwards will either have to be pretty loose, or be expected to fail frequently... But I haven't seen the validation, so maybe 7 angstroms is simultaneously tight enough to be informative, and loose enough not to cause spurious assertion errors in any of the practical cases tested...)

Reply:
I'm not sure tbh - I think we can add this as an objective function but we'd need to figure out the penalty for the RMSD here. For complex/solvent minimization, the outer shell of the environment is frozen, so it'd be very unlikely to escape too much. 7 angstroms seems to have passed on the hif2a set - and any atom that moves further than that will be problematic

Reply:
Hmm, fair enough -- I'm not sure what cases this assertion will catch / what the ideal intended behavior of the minimizer should be, but (1) picking a loose threshold like this seems safe as a post-condition, (2) seems somewhat safe to assume unrestrained minimization will introduce no other unwanted changes to the conformer...

(Although we may want to revisit (2) in the future to be more explicit, e.g. to prevent the minimizer from flipping important torsions, or to prevent the minimizer from moving any atom by more than 7 angstroms.)